coordinatorlib/src/main/scala/com/socrata/datacoordinator/secondary/sql/PostgresSecondaryMetrics.scala

```diff
@@ -0,0 +1,46 @@
+package com.socrata.datacoordinator.secondary.sql
+
+import java.sql.Connection
+
+import com.rojoma.simplearm.v2.using
+import com.socrata.datacoordinator.id.DatasetId
+import com.socrata.datacoordinator.id.sql._
+import com.socrata.datacoordinator.secondary.SecondaryMetric
+import com.socrata.datacoordinator.util.PostgresUniqueViolation
+
+class PostgresSecondaryMetrics(conn: Connection) extends SqlSecondaryMetrics(conn) {
+
+
+  private val log = org.slf4j.LoggerFactory.getLogger(classOf[PostgresSecondaryMetrics])
+
+  override def upsertDataset(storeId: String, datasetId: DatasetId, metric: SecondaryMetric): Unit = {
+    val rowsUpdated = using(conn.prepareStatement(
+      """UPDATE secondary_metrics
+        |   SET total_size = ?, updated_at = now()
+        | WHERE store_id = ?
+        |   AND dataset_system_id = ?""".stripMargin)) { stmt =>
+      stmt.setLong(1, metric.totalSizeBytes)
+      stmt.setString(2, storeId)
+      stmt.setDatasetId(3, datasetId)
+      stmt.executeUpdate()
+    }
+
+    // TODO: replace this with using Postgres UPSERT functionality introduced in 9.5
+    // once we are no longer running 9.4 truth instances
+    if (rowsUpdated == 0) {
+      try {
+        using(conn.prepareStatement("INSERT INTO secondary_metrics (store_id, dataset_system_id, total_size) (?, ?, ?)")) { stmt =>
+          stmt.setString(1, storeId)
+          stmt.setDatasetId(2, datasetId)
+          stmt.setLong(3, metric.totalSizeBytes)
+          stmt.execute()
+        }
+      } catch {
+        case uv@PostgresUniqueViolation(_*) =>
+          // there should only be one thread trying to due this at a time since this is only called in the SecondaryWatcher
```

Comment:
Technically this isn't true in the current codepath is it since we release the secondary_manifest lock before we grab the stats?

Reply:
You are right, ignore that.  I was getting confused with the resync lock.

![bitmoji](https://render.bitstrips.com/v2/cpanel/4a148723-fbbf-4972-8e6e-120148c5a497-69d0c8e1-e986-47fb-a727-1a8ae0328e06-v1.png?transparent=1&palette=1&width=246)

Reply:
```
LOG:  execute <unnamed>: SHOW TRANSACTION ISOLATION LEVEL
LOG:  execute <unnamed>: SET SESSION CHARACTERISTICS AS TRANSACTION ISOLATION LEVEL READ COMMITTED
LOG:  execute <unnamed>: SHOW TRANSACTION ISOLATION LEVEL
LOG:  execute S_2: BEGIN
LOG:  execute <unnamed>: UPDATE secondary_manifest
	SET latest_secondary_data_version = $1
	  ,cookie = $2
	  ,went_out_of_sync_at = CURRENT_TIMESTAMP
	WHERE claimant_id = $3
	  AND store_id = $4
	  AND dataset_system_id = $5
DETAIL:  parameters: $1 = '4', $2 = NULL, $3 = '61e9a209-98e7-4daa-9c43-5778a96e1d8a', $4 = 'pg', $5 = '2'
LOG:  execute S_1: COMMIT
LOG:  execute <unnamed>: SHOW TRANSACTION ISOLATION LEVEL
LOG:  execute S_2: BEGIN
LOG:  execute <unnamed>: DELETE FROM resync WHERE dataset_system_id = $1 AND store_id = $2 AND group_name = $3
DETAIL:  parameters: $1 = '2', $2 = 'pg', $3 = 'read'
LOG:  execute <unnamed>: BEGIN
LOG:  execute <unnamed>: SELECT dataset_system_id, disabled
	  FROM dataset_internal_name_map
	 WHERE dataset_internal_name = $1

DETAIL:  parameters: $1 = 'primus.2'
LOG:  execute <unnamed>: SELECT sum(pg_total_relation_size(oid)) AS total_size_bytes
	  FROM pg_class
	  WHERE relkind = 'r'
	    AND relname LIKE $1

DETAIL:  parameters: $1 = 't2\_%'
LOG:  execute S_2: COMMIT
LOG:  execute <unnamed>: SELECT 1
LOG:  execute <unnamed>: UPDATE secondary_metrics
	   SET total_size = $1, updated_at = now()
	 WHERE store_id = $2
	   AND dataset_system_id = $3
DETAIL:  parameters: $1 = '163840', $2 = 'pg', $3 = '2'
LOG:  execute <unnamed>: SELECT * FROM secondary_move_jobs WHERE dataset_system_id = $1 AND to_store_id = $2 AND (move_from_store_completed_at IS NULL OR move_to_store_completed_at IS NULL)
DETAIL:  parameters: $1 = '2', $2 = 'pg'
LOG:  execute <unnamed>: UPDATE secondary_move_jobs SET move_to_store_completed_at = now() WHERE dataset_system_id = $1 AND to_store_id = $2 AND move_to_store_completed_at IS NULL
DETAIL:  parameters: $1 = '2', $2 = 'pg'
LOG:  execute <unnamed>: SAVEPOINT JDBC_SAVEPOINT_1
LOG:  execute <unnamed>: UPDATE secondary_manifest
	SET claimed_at = NULL
	  ,claimant_id = NULL
	WHERE claimant_id = $1
	  AND store_id = $2
	  AND dataset_system_id = $3
DETAIL:  parameters: $1 = '61e9a209-98e7-4daa-9c43-5778a96e1d8a', $2 = 'pg', $3 = '2'
``` Logs to prove it :P